### PR TITLE
Fix rebase issuse

### PR DESCRIPTION
**Pull request by quintinwang5**
llvm branch: https://github.com/quintinwang5/llvm-project/tree/triton_rebase_b1115f8c

### CODE REVIEW

**Comment by chuanqi129** [comment]
@quintinwang5 Please try to use those llvm pre-built packages in https://github.com/intel/intel-xpu-backend-for-triton/releases/tag/llvm_pkg_b1115f8c to update the cmakelist.txt and verify them whether work fine.

**Comment by chuanqi129** [comment]
the UT test failed https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/6636312972/job/18028584890?pr=118 seems cause by the runner don't have basekit 2023.2 version now, need to check it @ESI-SYD 
```
/home/sdp/env_triton.sh: line 3: /home/sdp/intel/oneapi/compiler/2023.2.0/env/vars.sh: No such file or directory
/home/sdp/env_triton.sh: line 5: /home/sdp/intel/oneapi/mkl/2023.2.0/env/vars.sh: No such file or directory
```

And the inductor model level e2e test has more accuracy failure https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/6636312968/job/18028584208?pr=118 @quintinwang5 
```
============ Acc Check for HF amp_bf16 ============
num_total_amp_bf16: 46
num_passed_amp_bf16_inf: 39
num_failed_amp_bf16_inf: 7
amp_bf16_inf_acc_pass_rate: 84.78%
============ Acc Check for HF amp_fp16 ============
num_total_amp_fp16: 46
num_passed_amp_fp16_inf: 39
num_failed_amp_fp16_inf: 7
amp_fp16_inf_acc_pass_rate: 84.78%
```

**Comment by quintinwang5** [comment]
> the UT test failed https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/6636312972/job/18028584890?pr=118 seems cause by the runner don't have basekit 2023.2 version now, need to check it @ESI-SYD
> 
> ```
> /home/sdp/env_triton.sh: line 3: /home/sdp/intel/oneapi/compiler/2023.2.0/env/vars.sh: No such file or directory
> /home/sdp/env_triton.sh: line 5: /home/sdp/intel/oneapi/mkl/2023.2.0/env/vars.sh: No such file or directory
> ```
> 
> And the inductor model level e2e test has more accuracy failure https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/6636312968/job/18028584208?pr=118 @quintinwang5
> 
> ```
> ============ Acc Check for HF amp_bf16 ============
> num_total_amp_bf16: 46
> num_passed_amp_bf16_inf: 39
> num_failed_amp_bf16_inf: 7
> amp_bf16_inf_acc_pass_rate: 84.78%
> ============ Acc Check for HF amp_fp16 ============
> num_total_amp_fp16: 46
> num_passed_amp_fp16_inf: 39
> num_failed_amp_fp16_inf: 7
> amp_fp16_inf_acc_pass_rate: 84.78%
> ```



> the UT test failed https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/6636312972/job/18028584890?pr=118 seems cause by the runner don't have basekit 2023.2 version now, need to check it @ESI-SYD
> 
> ```
> /home/sdp/env_triton.sh: line 3: /home/sdp/intel/oneapi/compiler/2023.2.0/env/vars.sh: No such file or directory
> /home/sdp/env_triton.sh: line 5: /home/sdp/intel/oneapi/mkl/2023.2.0/env/vars.sh: No such file or directory
> ```
> 
> And the inductor model level e2e test has more accuracy failure https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/6636312968/job/18028584208?pr=118 @quintinwang5
> 
> ```
> ============ Acc Check for HF amp_bf16 ============
> num_total_amp_bf16: 46
> num_passed_amp_bf16_inf: 39
> num_failed_amp_bf16_inf: 7
> amp_bf16_inf_acc_pass_rate: 84.78%
> ============ Acc Check for HF amp_fp16 ============
> num_total_amp_fp16: 46
> num_passed_amp_fp16_inf: 39
> num_failed_amp_fp16_inf: 7
> amp_fp16_inf_acc_pass_rate: 84.78%
> ```

@chuanqi129 We should update pytorch and IPEX. These problems are caused by frontend codegen before triton.

**Comment by ESI-SYD** [comment]
> the UT test failed https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/6636312972/job/18028584890?pr=118 seems cause by the runner don't have basekit 2023.2 version now, need to check it @ESI-SYD
> 
> ```
> /home/sdp/env_triton.sh: line 3: /home/sdp/intel/oneapi/compiler/2023.2.0/env/vars.sh: No such file or directory
> /home/sdp/env_triton.sh: line 5: /home/sdp/intel/oneapi/mkl/2023.2.0/env/vars.sh: No such file or directory
> ```
> 

Reinstall Basekit 2023.2.0 on this runner

**Comment by quintinwang5** [comment]
@chuanqi129 @ESI-SYD 
pytorch： [pytorch-2.1](https://github.com/intel-innersource/frameworks.ai.pytorch.private-gpu/tree/pytorch-2.1)
IPEX: [master](https://github.com/intel-innersource/frameworks.ai.pytorch.ipex-gpu/commit/7980a37028023037b4f0b47617c5fc3343a6d09b)

Verified and passed model failures with oneapi 2024.0.

File: python/__init__.py [hunk 1]
```diff
@@ -10,15 +10,44 @@
...
+    with open(os.path.join(TRITON_PATH, "_C/libtriton.so"), "rb") as f:
+        contents += [hashlib.md5(f.read()).hexdigest()]
+    # language
+    language_path = os.path.join(TRITON_PATH, 'language')
+    for lib in pkgutil.iter_modules([language_path]):
+        with open(lib.module_finder.find_spec(lib.name).origin, "rb") as f:
+            contents += [hashlib.md5(f.read()).hexdigest()]
+    # ptxas version
+    try:
+        ptxas_version = hashlib.md5(subprocess.check_output(["ptxas", "--version"])).hexdigest()
```
Comment (chengjunlu) [inline python/__init__.py:45]: We don't use the ptxas tools. We can skip this.

File: python/__init__.py [hunk 2]
```diff
@@ -10,15 +10,44 @@
...
+    # frontend
+    with open(__file__, "rb") as f:
+        contents += [hashlib.md5(f.read()).hexdigest()]
+    # compiler
+    compiler_path = os.path.join(TRITON_PATH, 'compiler')
+    for lib in pkgutil.iter_modules([compiler_path]):
+        with open(lib.module_finder.find_spec(lib.name).origin, "rb") as f:
+            contents += [hashlib.md5(f.read()).hexdigest()]
+    # backend
+    with open(os.path.join(TRITON_PATH, "_C/libtriton.so"), "rb") as f:
```
Comment (chengjunlu) [inline python/__init__.py:36]: We uses the libxpu_intel_backend.so. Add that library as part of the hash key.

**Review by chengjunlu — APPROVED** [review]

**Comment by chuanqi129** [comment]
> @chuanqi129 @ESI-SYD pytorch： [pytorch-2.1](https://github.com/intel-innersource/frameworks.ai.pytorch.private-gpu/tree/pytorch-2.1) IPEX: [master](https://github.com/intel-innersource/frameworks.ai.pytorch.ipex-gpu/commit/7980a37028023037b4f0b47617c5fc3343a6d09b)
> 
> Verified and passed model failures with oneapi 2024.0.

Hi @ESI-SYD let's update the ci/nightly with latest IPEX 2.1 code base

**Review by chuanqi129 — APPROVED** [review]

File: .github/scripts/env_prepare.sh [hunk 3]
```diff
@@ -1,11 +1,11 @@
 JOB_WORKSPACE=${1:-triton-preci}
 TORCH_REPO=${2:-https://github.com/pytorch/pytorch.git}
-TORCH_BRANCH=${3:-v2.0.1}
-TORCH_COMMIT=${4:-e9ebda29d87ce0916ab08c06ab26fd3766a870e5}
+TORCH_BRANCH=${3:-release/2.1}
```
Comment (chuanqi129) [inline .github/scripts/env_prepare.sh:3]: does this branch based on pytorch v2.1?

Comment (quintinwang5) [reply .github/scripts/env_prepare.sh:3]: Yes. But I didn't find public branches which match with the verified private branches.

Comment (ESI-SYD) [reply .github/scripts/env_prepare.sh:3]: release/2.1 branches raised `RuntimeError: PyTorch is not linked with support for xpu devices`, https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/6663470255/job/18109403706?pr=118

**Comment by chuanqi129** [comment]
Please also change the default oneapi version to `2024.0` in https://github.com/intel/intel-xpu-backend-for-triton/blob/main/.github/scripts/env_prepare.sh#L8 & https://github.com/intel/intel-xpu-backend-for-triton/blob/main/.github/scripts/env_triton.sh#L1

**Comment by ESI-SYD** [comment]
Error found when ipex import in CI checks  `ImportError: /home/gta/intel/oneapi/mkl/2024.0/lib/libmkl_gnu_thread.so.2: undefined symbol: mkl_blas_xgemm_f16f16f32_pack` [E2E](https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/6708123140/job/18255217396?pr=118)

**Comment by quintinwang5** [comment]
> Error found when ipex import in CI checks `ImportError: /home/gta/intel/oneapi/mkl/2024.0/lib/libmkl_gnu_thread.so.2: undefined symbol: mkl_blas_xgemm_f16f16f32_pack` [E2E](https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/6708123140/job/18255217396?pr=118)

@ESI-SYD Seems it's an environment problem on preCI machine. And I can't reproduce this problem locally. Could you help to check whether if these necessary libraries are in correct path. (The undefined symbol should be in `libmkl_core.so.2` under the same directory.

**Comment by quintinwang5** [comment]
E2E model new fail: 1
```
num_total_amp_bf16: 46
num_passed_amp_bf16_inf: 45
num_failed_amp_bf16_inf: 1
amp_bf16_inf_acc_pass_rate: 97.83%
```
xpu,GPT2ForSequenceClassification,1,fail_accuracy,644,2,1,1
UT new fail: 63
```
============ UT Status Overview ============
num_total: 6716
num_skipped: 1157
num_failed: 1[27](https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/6741390098/job/18325910412?pr=118#step:9:28)
num_passed: 5432
total_pass_rate: 97.72%
```
```
test_core.py::test_scan2d[get_first_element-int32-shape29-0-4]
test_core.py::test_reduce[1-sum-float32-shape117-0]
test_core.py::test_scan2d[get_first_element-float32-shape146-1-16]
test_core.py::test_scan2d[get_first_element-int32-shape95-1-16]
test_core.py::test_scan2d[get_first_element-float32-shape77-0-4]
test_core.py::test_scan2d[get_first_element-float32-shape164-0-16]
test_core.py::test_scan2d[get_first_element-int32-shape32-0-4]
test_core.py::test_scan2d[get_first_element-float32-shape59-1-4]
test_core.py::test_scan2d[get_first_element-int32-shape35-0-4]
test_core.py::test_scan2d[get_first_element-float32-shape62-1-4]
test_core.py::test_scan2d[get_first_element-int32-shape38-0-4]
test_core.py::test_scan2d[get_first_element-float32-shape149-0-16]
test_core.py::test_scan2d[get_first_element-float32-shape80-0-4]
test_core.py::test_scan2d[get_first_element-int32-shape101-1-16]
test_core.py::test_scan2d[get_first_element-int32-shape104-1-16]
test_core.py::test_scan2d[get_first_element-int32-shape107-0-16]
test_core.py::test_scan2d[get_first_element-int32-shape2-1-4]
test_core.py::test_scan2d[get_first_element-float32-shape65-0-4]
test_core.py::test_scan2d[get_first_element-float32-shape44-1-4]
test_core.py::test_scan2d[get_first_element-float32-shape152-0-16]
test_core.py::test_scan2d[get_first_element-int32-shape110-0-16]
test_core.py::test_scan2d[get_first_element-float32-shape155-0-16]
test_core.py::test_scan2d[get_first_element-int32-shape[86](https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/6741390098/job/18325910412?pr=118#step:11:87)-1-16]
test_core.py::test_scan2d[get_first_element-float32-shape158-0-16]
test_core.py::test_scan2d[get_first_element-float32-shape161-0-16]
test_core.py::test_scan2d[get_first_element-int32-shape5-1-4]
test_core.py::test_scan2d[get_first_element-float32-shape47-1-4]
test_core.py::test_scan2d[get_first_element-float32-shape68-0-4]
test_core.py::test_scan2d[get_first_element-float32-shape71-0-4]
test_core.py::test_scan2d[get_first_element-float32-shape74-0-4]
test_core.py::test_scan2d[get_first_element-int32-shape[89](https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/6741390098/job/18325910412?pr=118#step:11:90)-1-16]
test_core.py::test_scan2d[get_first_element-int32-shape113-0-16]
test_core.py::test_scan2d[get_first_element-int32-shape116-0-16]
test_core.py::test_scan2d[get_first_element-int32-shape119-0-16]
test_core.py::test_scan2d[get_first_element-int32-shape122-0-16]
test_core.py::test_scan2d[get_first_element-int32-shape8-1-4]
test_core.py::test_scan2d[get_first_element-float32-shape50-1-4]
test_core.py::test_scan2d[get_first_element-float32-shape128-1-16]
test_core.py::test_scan2d[get_first_element-float32-shape131-1-16]
test_core.py::test_scan2d[get_first_element-int32-shape11-1-4]
test_core.py::test_scan2d[get_first_element-float32-shape53-1-4]
test_core.py::test_scan2d[get_first_element-float32-shape134-1-16]
test_core.py::test_scan2d[get_first_element-float32-shape137-1-16]
test_core.py::test_scan2d[get_first_element-int32-shape17-1-4]
test_core.py::test_scan2d[get_first_element-int32-shape20-1-4]
test_core.py::test_scan2d[get_first_element-float32-shape143-1-16]
test_core.py::test_scan2d[get_first_element-int32-shape23-0-4]
test_core.py::test_scan2d[get_first_element-int32-shape[92](https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/6741390098/job/18325910412?pr=118#step:11:93)-1-16]
test_core.py::test_scan2d[get_first_element-int32-shape26-0-4]
test_core.py::test_reduce[1-argmin-float32-shape123-0]
test_core.py::test_reduce[1-min-float32-shape[96](https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/6741390098/job/18325910412?pr=118#step:11:97)-0]
test_core.py::test_reduce[1-argmin-float32-shape126-0]
test_core.py::test_reduce[1-argmax-float32-shape132-0]
test_core.py::test_reduce[1-argmax-float32-shape135-0]
test_core.py::test_reduce[1-min-float32-shape[99](https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/6741390098/job/18325910412?pr=118#step:11:100)-0]
test_core.py::test_reduce[1-max-float32-shape[105](https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/6741390098/job/18325910412?pr=118#step:11:106)-0]
test_core.py::test_reduce[1-max-float32-shape[108](https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/6741390098/job/18325910412?pr=118#step:11:109)-0]
test_core.py::test_reduce[1-sum-float32-shape[114](https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/6741390098/job/18325910412?pr=118#step:11:115)-0]
test_core.py::test_constexpr_propagation
test_core.py::test_enable_fp_fusion[False]
test_core.py::test_enable_fp_fusion[True]
test_subprocess.py::test_assert[double_assert]
============ New Failed Not TODO Cases ============
test_core.py::test_constexpr_propagation
test_subprocess.py::test_assert[double_assert]
test_core.py::test_constexpr_propagation
test_subprocess.py::test_assert[double_assert]
num_new_failed_cases: 62
num_new_failed_not_todo_cases: 2
```

**Comment by ESI-SYD** [comment]
Add E2E CI for **amp_fp16** acc check result, Inductor E2E CI unexpected exit while print this dtype, For amp_bf16: **GPT2ForSequenceClassification** fails is a known issue
That is to say E2E CI PASS in this PR.
```
num_total_amp_fp16: 46
num_passed_amp_fp16_inf: 46
num_failed_amp_fp16_inf: 0
amp_fp16_inf_acc_pass_rate: 100.00%
```

**Comment by quintinwang5** [comment]
> Add E2E CI for **amp_fp16** acc check result, Inductor E2E CI unexpected exit while print this dtype, For amp_bf16: **GPT2ForSequenceClassification** fails is a known issue That is to say E2E CI PASS in this PR.
> 
> ```
> num_total_amp_fp16: 46
> num_passed_amp_fp16_inf: 46
> num_failed_amp_fp16_inf: 0
> amp_fp16_inf_acc_pass_rate: 100.00%
> ```

UT failures are led by updates in `test_core.py`. WIP to fix them.  Then, should we set these known model/UT failures to expected and move forward or wait for their addressments? @EikanWang @chengjunlu @chuanqi129

**Comment by chuanqi129** [comment]
> 
> UT failures are led by updates in `test_core.py`. WIP to fix them. Then, should we set these known model/UT failures to expected and move forward or wait for their addressments? @EikanWang @chengjunlu @chuanqi129 @riverliuintel

We also can move those failure case into https://github.com/intel/intel-xpu-backend-for-triton/blob/main/.github/tests/triton_todo_failure_tests.log in this PR, and move the triton commit forward. How do you think?

**Comment by ESI-SYD** [comment]
```
============ Acc Check for HF amp_bf16 ============
num_total_amp_bf16: 46
num_passed_amp_bf16_inf: 45
num_failed_amp_bf16_inf: 1
amp_bf16_inf_acc_pass_rate: 97.83%
============ Acc Check for HF amp_fp16 ============
num_total_amp_fp16: 46
num_passed_amp_fp16_inf: 46
num_failed_amp_fp16_inf: 0
amp_fp16_inf_acc_pass_rate: 100.00%


```

**Comment by chengjunlu** [comment]
> ```
> ============ Acc Check for HF amp_bf16 ============
> num_total_amp_bf16: 46
> num_passed_amp_bf16_inf: 45
> num_failed_amp_bf16_inf: 1
> amp_bf16_inf_acc_pass_rate: 97.83%
> ============ Acc Check for HF amp_fp16 ============
> num_total_amp_fp16: 46
> num_passed_amp_fp16_inf: 46
> num_failed_amp_fp16_inf: 0
> amp_fp16_inf_acc_pass_rate: 100.00%
> ```

Let's create a task to track the random BF16 model accuracy issue. And I will merge the rebasing PR which is not the root cause of the random accuracy issue.

**Pull request merged**